aircmd/actions/environments.py

```diff
@@ -12,30 +12,31 @@
 
 from dagger import CacheSharingMode, CacheVolume, Container, Directory, File
 
-from ..models.base import GlobalSettings, PipelineContext
+from ..models.base import GlobalSettings, Pipeline
 from .pipelines import get_file_contents, get_repo_dir, with_exit_code
 from .strings import slugify
 
 
-def with_python_base(context: PipelineContext, python_image_name: str = "python:3.9-slim") -> Container:
+def with_python_base(pipeline: Pipeline, python_image_name: str = "python:3.11-slim") -> Container:
     """Build a Python container with a cache volume for pip cache.
-
+    
     Args:
-        context (PipelineContext): The current test context, providing a dagger client and a repository directory.
-        python_image_name (str, optional): The python image to use to build the python base environment. Defaults to "python:3.9-slim".
+        context (Pipeline): The current test pipeline, providing a dagger client and a repository directory.
+        python_image_name (str, optional): The python image to use to build the python base environment. Defaults to "python:3.11-slim".
 
     Raises:
         ValueError: Raised if the python_image_name is not a python image.
 
     Returns:
         Container: The python base environment container.
     """
+    
     if not python_image_name.startswith("python:3"):
```

Comment:
There is nothing stopping someone from including their own base image, they just have to declare it in their pipeline:
```python
        client.container()
        .from_("python:alpine")
        .with_exec(["pip", "install", "cowsay"])
        .with_entrypoint(["cowsay"])
        .with_exec([msg])
        .stdout()
  ```

Maybe we should add some notes at the top of this file, this is intended to be a pretty close 1:1 analogue for how docker image layers work. A library of commonly used ones that can be shared across pipelines so that they can be leveraged for caching.

The pattern is ideally that someone writing their own plugin reaches for this library of layers first to leverage the shared caching across everything and reduction of boilerplate, but if you need something custom you can either just inline your own image like in my above example or create your own custom library of "environments" that extends this one